google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerException.java

```diff
@@ -59,10 +63,23 @@ public String getResourceName() {
       boolean retryable,
       @Nullable String message,
       @Nullable Throwable cause) {
-    super(message, cause, code.getCode(), retryable);
+    super(
+        message,
+        // If the cause is instance of APIException then using its cause to avoid the breaking
+        // change, because earlier we were passing APIException's cause to constructor.
+        cause instanceof ApiException ? cause.getCause() : cause,
```

Comment:
Hmm... Now that I see it a second time, I think we need to add a constructor overload that takes the `ApiException` as a separate argument, after all. I think that will make it a little more readable and understandable what is going on, but more importantly: The current implementation might also break existing code. Anyone that called this constructor directly with an instance of `ApiException` the cause (so not through the specific method in `SpannerExceptionFactory`) will now receive a different exception, as the `ApiException` is no longer set as the cause of this exception.

Reply:
Yeah, I think having the parameter based constructor will be more readable and understandable. SpannerException constructor is package-private and because we have factory written over it, in an ideal world everyone should use it to create the SpannerException, but as you correctly mentioned we can go ahead with parameterised constructor because on the top of readability and understandability we will also have more granular control to create SpannerException.